### PR TITLE
[WIP] Remove catalog activations locking

### PR DESCRIPTION
**Pull request by ReubenBond**

### CODE REVIEW

**Comment by ReubenBond** [comment]
@dotnet-bot test bvt

known flaky test

File: src/Orleans.Runtime/Catalog/Catalog.cs [hunk 1]
```diff
@@ -443,20 +439,19 @@ public void GetGrainTypeInfo(int typeCode, out string grainClass, out PlacementS
             activatedPromise = Task.CompletedTask;
             PlacementStrategy placement;
 
-            lock (activations)
+            while (true)
```
Comment (xiazen) [inline src/Orleans.Runtime/Catalog/Catalog.cs]: this smells bad. why are we adding infinite retry here while the original logic didn't?

**Comment by jdom** [comment]
>                     // We want to do this (RegisterMessageTarget) under the same lock that we tested TryGetActivationData. They both access ActivationDirectory.

no longer true

---
Refers to: src/Orleans.Runtime/Catalog/Catalog.cs:474 in cfd0c4b. [](commit_id = cfd0c4bf8d2f74abfb450b3db1c0af5d39fa6257, deletion_comment = False)

**Comment by jdom** [comment]
>         private void OnSiloStatusChange(SiloAddress updatedSilo, SiloStatus status)

can this be called concurrently or is this single threaded? Because if it's the first option, you'll have to use some kind of locking primitive, not necessarily on `activations`.

---
Refers to: src/Orleans.Runtime/Catalog/Catalog.cs:1383 in cfd0c4b. [](commit_id = cfd0c4bf8d2f74abfb450b3db1c0af5d39fa6257, deletion_comment = False)

File: src/Orleans.Runtime/Catalog/Catalog.cs (hunk 1 above)
Comment (jdom) [reply src/Orleans.Runtime/Catalog/Catalog.cs]: this is a standard lock-free approach when dealing with optimistic concurrency conflicts

---
In reply to: [274565380](https://github.com/dotnet/orleans/pull/5526#discussion_r274565380) [](ancestors = 274565380)

Comment (xiazen) [reply src/Orleans.Runtime/Catalog/Catalog.cs]: I see. Thanks for explaining !

**Pull request closed without merging**